### PR TITLE
Crudl

### PR DESCRIPTION
**Pull request by jcassem**
- Delete entities and re-create it if the database instance is Cosmos DB due to a limitation with the gremlin lib we used (details in README).
- Automatically set Id to UUID if not set

### CODE REVIEW

File: src/main/java/com/example/entityapi/EntityApiApplication.java [hunk 1]
```diff
@@ -14,21 +13,29 @@
...
-public class EntityApiApplication {
-	public static void main(String[] args) {
-		SpringApplication.run(EntityApiApplication.class, args);
+public class EntityApiApplication
+{
+	public static void main(String[] args)
+	{
+		SpringApplication.run(EntityApiApplication.class,
+							  args
+							 );
```
Comment (RichTeaMan) [inline src/main/java/com/example/entityapi/EntityApiApplication.java]: Your formatter is the Pizza Hut of formatters.

File: src/main/java/com/example/entityapi/service/EntityService.java [hunk 2]
```diff
@@ -51,7 +57,34 @@ public Entity get(String id)
...
+		var isUpdate = entity.getId() != null && !entity.getId().isBlank();
+		if (!isUpdate)
+		{
+			entity.setId(UUID.randomUUID().toString());
+		}
+		else
+		{
+			// entity has to be deleted and re-created due to this gremlin
+			// libraries incompatibility with Cosmos DB partition keys.
+			if (isCosmosDb)
```
Comment (RichTeaMan) [inline src/main/java/com/example/entityapi/service/EntityService.java]: else if

File: src/main/java/com/example/entityapi/service/EntityService.java [hunk 3]
```diff
@@ -51,7 +57,34 @@ public Entity get(String id)
...
+
 		logger.debug("Create/update entity: " + entity.getId());
+
+		if (entity.getComments() != null)
+		{
+			for (var comment : entity.getComments())
+			{
+				if (comment.getId() == null || comment.getId().isBlank())
+				{
+					comment.setId(UUID.randomUUID().toString());
```
Comment (RichTeaMan) [inline src/main/java/com/example/entityapi/service/EntityService.java]: A stream lamdba would make less work of this.

File: src/main/java/com/example/entityapi/config/CorsConfiguration.java [hunk 4]
```diff
@@ -10,9 +10,14 @@
...
-				.allowedMethods("GET", "POST", "PUT", "DELETE");
+	public void addCorsMappings(CorsRegistry registry)
+	{
+		registry.addMapping("/**")
+				.allowedMethods("GET",
+								"POST",
+								"PUT",
+								"DELETE",
+								"PATCH"
+							   );
```
Comment (RichTeaMan) [inline src/main/java/com/example/entityapi/config/CorsConfiguration.java]: I can't stop you merging, but I've run devs out of town for less.

File: src/main/java/com/example/entityapi/EntityApiApplication.java [hunk 5]
```diff
@@ -14,21 +13,29 @@
...
 	 * Creates object mapper with zalando problem modules.
 	 */
 	@Bean
-	public ObjectMapper objectMapper() {
-		return new ObjectMapper().registerModules(new ProblemModule(), new ConstraintViolationProblemModule());
+	public ObjectMapper objectMapper()
+	{
+		return new ObjectMapper().registerModules(new ProblemModule(),
+												  new ConstraintViolationProblemModule()
+												 );
```
Comment (RichTeaMan) [inline src/main/java/com/example/entityapi/EntityApiApplication.java]: Sir, has your formatter been drinking?

File: src/main/java/com/example/entityapi/controller/FileUploadController.java [hunk 6]
```diff
@@ -1,66 +1,77 @@
...
+		var data = storageService.downloadFile(filename);
+		// Set headers
+		final HttpHeaders headers = new HttpHeaders();
+		headers.setContentType(MediaType.parseMediaType(URLConnection.guessContentTypeFromName(filename)));
 
-        var entity = new ResponseEntity<byte[]>(data, headers, HttpStatus.OK);
+		var entity = new ResponseEntity<byte[]>(data,
+												headers,
+												HttpStatus.OK
+		);
```
Comment (RichTeaMan) [inline src/main/java/com/example/entityapi/controller/FileUploadController.java]: Jeez, that contructor had a family.

File: src/main/java/com/example/entityapi/service/EntityLinkService.java [hunk 7]
```diff
@@ -51,6 +52,11 @@ public EntityLink get(String id)
 	@Override
 	public EntityLink createUpdate(EntityLink entityLink)
 	{
+		if (entityLink.getId() == null || entityLink.getId().isBlank())
```
Comment (RichTeaMan) [inline src/main/java/com/example/entityapi/service/EntityLinkService.java]: StingUtils.isBlank could be the play.

File: src/main/java/com/example/entityapi/service/EntityLinkService.java [hunk 8]
```diff
@@ -51,6 +52,11 @@ public EntityLink get(String id)
...
+		{
+			entityLink.setId(UUID.randomUUID().toString());
+		}
+
 		logger.debug("Create/update entity link: " + entityLink.getId());
```
Comment (RichTeaMan) [inline src/main/java/com/example/entityapi/service/EntityLinkService.java]: Ideally the log would say which of a create or update occurred.

File: README.md [hunk 9]
```diff
@@ -76,6 +76,22 @@ Further information about the Gremlin API and this setup can be found on this
...
+
+For Cosmos DB instances you will also need to set the following property in your `application.property`
+so this service knows to delete and re-create entities rather than attempt to update them due to 
+Cosmos DB throwing an error when you attempt to amend the databases partition key, even with the same value.
+This is due to the fact that partition keys are not a concern Gremlin, but are for Cosmos DB, therefore the 
+underlying library used here (as mentioned above) has some mild incompatibility with Cosmos DB.
+
+Property:
+```properties
+database.isComsosDb=true
```
Comment (RichTeaMan) [inline README.md]: coMsos

File: src/main/java/com/example/entityapi/service/EntityLinkService.java (hunk 8 above)
Comment (jcassem) [reply src/main/java/com/example/entityapi/service/EntityLinkService.java]: The controller layer will do this.

File: src/main/java/com/example/entityapi/EntityApiApplication.java (hunk 1 above)
Comment (jcassem) [reply src/main/java/com/example/entityapi/EntityApiApplication.java]: eeesh thats pretty bad.

**Pull request merged**